syntax = "proto3";

package master;

service MasterService {
    rpc create (Location) returns (FileMapping);
    rpc mkdir (Location) returns (FileMapping);

    rpc read (ReadRequest) returns (FileMapping);
    rpc write (WriteRequest) returns (BlockList);
    rpc delete (DeleteRequest) returns (FileMapping);
    rpc getFileTableEntry (GetFileTableEntryRequest) returns (FileMapping);
    rpc getListOfFiles (Location) returns (FileList);
    rpc getBlockSize (Empty) returns (BlockSize);
    rpc getMinions (Empty) returns (MinionList);
}

message Location {
    string path = 1;
    oneof extras {
        int32 mode = 2;
    }
}


message FileMapping {
    // attrs optional?
    repeated Block blocks = 1;
    oneof extras {
        string attrs = 2;
    }
}

message Block {
    string block_uuid = 1;
    string node_id = 2;
    int32 block_index = 3;
}

message WriteRequest {
    string dest = 1;
    int32 size = 2;
}

message BlockList {
    repeated Block blocks = 1;
}

message ReadRequest {
    string fname = 1;
}

message DeleteRequest {
    string fname = 1;
}

message GetFileTableEntryRequest {
    string fname = 1;
}

message FileList {
    repeated string files = 1;
}

message BlockSize {
    int32 size = 1;
}

message MinionList {
    map<string, string> minions = 1;
}

message Empty {}
